### PR TITLE
Better throttling

### PR DESCRIPTION
**Pull request by hroff-1902**
Preparation for (part of) #2885 

Split from the upcoming rest to have it observable, for easy review.

* Worker now throttles (with an empty function) in the STOPPED state as well, instead of explicit sleep, this allows to print same diagnostics messages in the stopped state as well.
* Better throttling: the "====..." divider (at debug level) is printed before the working function, to efficiently separate startup messages in the log from the messages of the first throttling iteration
* The heartbeat message moved from FreqtradeBot into Worker and issued at the stopped state as well.
* The heartbeat message now contains the state, to see something when the bot was stopped.
* The heartbeat message now contains the version of the bot. I added this for us to see the version from the logs posted by users, which forget to fill the new issue form. However, I do not like it actually, the version info can be removed from the heartbeat message.

The overall look of the new heartbeat message:
```
...
2020-02-23 00:32:03,654 - freqtrade.exchange.exchange - INFO - Using Exchange "Binance"
2020-02-23 00:32:04,485 - freqtrade.resolvers.exchange_resolver - INFO - Using resolved exchange 'Binance'...
2020-02-23 00:32:04,505 - freqtrade.wallets - INFO - Wallets synced.
2020-02-23 00:32:04,508 - freqtrade.resolvers.iresolver - INFO - Using resolved pairlist StaticPairList from '/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/pairlist/StaticPairList.py'...
2020-02-23 00:32:04,508 - freqtrade.rpc.rpc_manager - INFO - Enabling rpc.api_server
2020-02-23 00:32:04,611 - freqtrade.rpc.api_server - INFO - Starting HTTP Server at 127.0.0.1:8080
2020-02-23 00:32:04,612 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'running'}
2020-02-23 00:32:04,612 - freqtrade.rpc.api_server - INFO - Starting Local Rest Server.
2020-02-23 00:32:04,612 - freqtrade.worker - INFO - Changing state to: RUNNING
2020-02-23 00:32:04,613 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': warning, 'status': 'Dry run is enabled. All trades are simulated.'}
2020-02-23 00:32:04,613 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': custom, 'status': "*Exchange:* `binance`\n*Stake per trade:* `0.05 BTC`\n*Minimum ROI:* `{'60': 0.01, '30': 0.02, '0': 0.04}`\n*Stoploss:* `-0.1`\n*Ticker Interval:* `1m`\n*Strategy:* `SampleStrategy`"}
2020-02-23 00:32:04,613 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': "Searching for BTC pairs to buy and sell based on [{'StaticPairList': 'StaticPairList'}]"}
2020-02-23 00:32:09,622 - freqtrade.worker - INFO - Bot heartbeat. PID=22502, version='develop-ca8e52dc', state='RUNNING'
2020-02-23 00:33:09,681 - freqtrade.worker - INFO - Bot heartbeat. PID=22502, version='develop-ca8e52dc', state='RUNNING'
2020-02-23 00:34:09,743 - freqtrade.worker - INFO - Bot heartbeat. PID=22502, version='develop-ca8e52dc', state='RUNNING'
```
then `/stop` command:
```
2020-02-23 00:34:55,791 - werkzeug - INFO - 127.0.0.1 - - [23/Feb/2020 00:34:55] "POST /api/v1/stop HTTP/1.1" 200 -
2020-02-23 00:34:59,786 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'stopped'}
2020-02-23 00:34:59,786 - freqtrade.worker - INFO - Changing state to: STOPPED
2020-02-23 00:35:04,791 - freqtrade.worker - INFO - Bot heartbeat. PID=22502, version='develop-ca8e52dc', state='STOPPED'
2020-02-23 00:36:04,846 - freqtrade.worker - INFO - Bot heartbeat. PID=22502, version='develop-ca8e52dc', state='STOPPED'
```
then `/reload_conf`:
```
2020-02-23 00:36:10,487 - werkzeug - INFO - 127.0.0.1 - - [23/Feb/2020 00:36:10] "POST /api/v1/reload_conf HTTP/1.1" 200 -
2020-02-23 00:36:14,853 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'reload_conf'}
2020-02-23 00:36:14,853 - freqtrade.worker - INFO - Changing state to: RELOAD_CONF
2020-02-23 00:36:14,854 - freqtrade.worker - INFO - Bot heartbeat. PID=22502, version='develop-ca8e52dc', state='RELOAD_CONF'
2020-02-23 00:36:14,854 - freqtrade.freqtradebot - INFO - Cleaning up modules ...
...
2020-02-23 00:36:15,027 - freqtrade.exchange.exchange - INFO - Using Exchange "Binance"
2020-02-23 00:36:15,915 - freqtrade.resolvers.exchange_resolver - INFO - Using resolved exchange 'Binance'...
2020-02-23 00:36:15,924 - freqtrade.wallets - INFO - Wallets synced.
2020-02-23 00:36:15,927 - freqtrade.resolvers.iresolver - INFO - Using resolved pairlist StaticPairList from '/home/user/freqtrade-wrk/github-hroff-1902/freqtrade/freqtrade/pairlist/StaticPairList.py'...
2020-02-23 00:36:15,927 - freqtrade.rpc.rpc_manager - INFO - Enabling rpc.api_server
2020-02-23 00:36:15,939 - freqtrade.rpc.api_server - INFO - Starting HTTP Server at 127.0.0.1:8080
2020-02-23 00:36:15,939 - freqtrade.rpc.api_server - INFO - Starting Local Rest Server.
2020-02-23 00:36:15,940 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'config reloaded'}
2020-02-23 00:36:15,940 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'running'}
2020-02-23 00:36:15,940 - freqtrade.worker - INFO - Changing state to: RUNNING
2020-02-23 00:36:15,940 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': warning, 'status': 'Dry run is enabled. All trades are simulated.'}
2020-02-23 00:36:15,940 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': custom, 'status': "*Exchange:* `binance`\n*Stake per trade:* `0.05 BTC`\n*Minimum ROI:* `{'60': 0.01, '30': 0.02, '0': 0.04}`\n*Stoploss:* `-0.1`\n*Ticker Interval:* `1m`\n*Strategy:* `SampleStrategy`"}
2020-02-23 00:36:15,940 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': "Searching for BTC pairs to buy and sell based on [{'StaticPairList': 'StaticPairList'}]"}
2020-02-23 00:36:20,945 - freqtrade.worker - INFO - Bot heartbeat. PID=22502, version='develop-ca8e52dc', state='RUNNING'
^C2020-02-23 00:36:33,455 - freqtrade.commands.trade_commands - INFO - SIGINT received, aborting ...
2020-02-23 00:36:33,456 - freqtrade.commands.trade_commands - INFO - worker found ... calling exit
2020-02-23 00:36:33,456 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'process died'}
2020-02-23 00:36:33,456 - freqtrade.freqtradebot - INFO - Cleaning up modules ...
2020-02-23 00:36:33,456 - freqtrade.rpc.rpc_manager - INFO - Cleaning up rpc modules ...
2020-02-23 00:36:33,456 - freqtrade.rpc.api_server - INFO - Stopping API Server
2020-02-23 00:36:33,465 - freqtrade.rpc.api_server - INFO - Local Rest Server started.
```

### CODE REVIEW

File: freqtrade/worker.py [hunk 1]
```diff
@@ -63,60 +67,74 @@ def run(self) -> None:
             if state == State.RELOAD_CONF:
                 self._reconfigure()
 
-    def _worker(self, old_state: Optional[State], throttle_secs: Optional[float] = None) -> State:
+    def _worker(self, old_state: Optional[State]) -> State:
         """
         Trading routine that must be run at each loop
```
Comment (xmatthias) [inline freqtrade/worker.py]: maybe reword this to 
```suggestion
        Main routine that must be run at each loop
```
?

File: freqtrade/worker.py [hunk 2]
```diff
@@ -63,60 +67,74 @@ def run(self) -> None:
...
+        logger.debug("========================================")
         result = func(*args, **kwargs)
-        end = time.time()
-        duration = max(min_secs - (end - start), 0.0)
-        logger.debug('Throttling %s for %.2f seconds', func.__name__, duration)
-        time.sleep(duration)
+        time_passed = time.time() - self.last_throttle_start_time
+        sleep_duration = max(throttle_secs - time_passed, 0.0)
+        logger.debug(f"Throttling with '{func.__name__}()': sleep for {sleep_duration:.2f} s, "
+                     f"last iteration took {time_passed:.2f} s.")
```
Comment (xmatthias) [inline freqtrade/worker.py]: wouldn't we want the "s" be right next to the number?

so "sleep for 2s" - instead of "sleep for 2 s"? or is it intentionally done this way?

File: freqtrade/worker.py [hunk 3]
```diff
@@ -26,12 +27,15 @@ def __init__(self, args: Dict[str, Any], config: Dict[str, Any] = None) -> None:
...
         """
-        logger.info('Starting worker %s', __version__)
+        logger.info(f"Starting worker {__version__}")
 
         self._args = args
         self._config = config
         self._init(False)
 
+        self.last_throttle_start_time: float = 0
+        self._heartbeat_msg: float = 0
```
Comment (xmatthias) [inline freqtrade/worker.py]: i know you only moved this here - but looking at the variable name now i did not really expect a time as content.

Maybe we can use `self._last_heartbeat_ts` instead ?

File: freqtrade/worker.py [hunk 4]
```diff
@@ -48,10 +52,10 @@ def _init(self, reconfig: bool) -> None:
...
-            'process_throttle_secs',
-            constants.PROCESS_THROTTLE_SECS
-        )
+        internals_config = self._config.get('internals', {})
+        self._throttle_secs = internals_config.get('process_throttle_secs',
+                                                   constants.PROCESS_THROTTLE_SECS)
+        self._heartbeat_interval = internals_config.get('heartbeat_interval', 60)
 
         self._sd_notify = sdnotify.SystemdNotifier() if \
             self._config.get('internals', {}).get('sd_notify', False) else None
```
Comment (xmatthias) [inline freqtrade/worker.py]: Since you've now got `internals_config` available directly i guess it could/should be used here too

File: freqtrade/worker.py (hunk 3 above)
Comment (hroff-1902) [reply freqtrade/worker.py]: Already done in the upcoming commits (not in this PR)

File: freqtrade/worker.py (hunk 4 above)
Comment (hroff-1902) [reply freqtrade/worker.py]: This also is done in the next episode of this saga... 😄

File: freqtrade/worker.py (hunk 2 above)
Comment (hroff-1902) [reply freqtrade/worker.py]: According to English typographical rules, as Wikipedia manuals say to us (and people at Wikipedia know this pretty well, believe me):

"Use a non-breaking space ({{nbsp}} or &nbsp;) between a number and a unit symbol, or use {{nowrap}", for program code/ASCII output it's blank space. 

Then, "Unit symbols are undotted". 

Btw, 'sec' is also allowed **unit symbol** for 'seconds' **unit name**, I wanted to use it as an alternative here (aligns better with 'min').

"Unit symbols (in any system) are identical in singular and plural." So, no 'secs', 'mins' are allowed, only 'seconds', 'minutes', 's (sec)', 'm (min)'.

See https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style/Dates_and_numbers#Unit_names_and_symbols.

**Comment by hroff-1902** [comment]
the logs above contain one unexpected behavior: it reports
```
freqtrade.worker - INFO - Bot heartbeat. PID=22502, version='develop-ca8e52dc', state='RELOAD_CONF'```
if you are on time... I did not expected this (only RUNNING and STOPPED), that's because of the commit where I moved the heartbeat to be issued just after the first throttling iteration after the change of state. The intent was not to wait for about a minute for 
```
freqtrade.worker - INFO - Bot heartbeat. PID=22502, version='develop-ca8e52dc', state='STOPPED'
```
after the `/stop` command and have it at the first iteration after that. It also caused this unexpected behavior, but I think it's okay...

(just logged it here)
